File: macos/Sources/ClaudeUsagePanel/SessionPing.swift
import ClaudeUsageCore
import Foundation

// Session pings from the app: reads and writes the same launchd agent as
// `./install.sh sessionping`, so the CLI and the Settings UI stay two
// frontends over one schedule. The plist on disk is the source of truth
// (like LoginItem's SMAppService state) - nothing is kept in UserDefaults.
enum SessionPing {
    static var plistURL: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/LaunchAgents/\(SessionPingAgent.label).plist")
    }

    struct State {
        var enabled: Bool
        var schedule: SessionPingSchedule
        var runner: String?
    }

    static func read() -> State {
        guard let data = try? Data(contentsOf: plistURL),
            let parsed = SessionPingAgent.parse(plistData: data)
        else {
            return State(enabled: false, schedule: .default, runner: nil)
        }
        return State(enabled: true, schedule: parsed.schedule, runner: parsed.runner)
    }

    /// The ping worker script. Prefer the path an existing agent already uses
    /// (a git-checkout install), else the copy bundled in the app's Resources
    /// by `install.sh macos`. A bare `swift build` binary has no bundle copy -
    /// the returned nil surfaces as an error line in Settings.
    static func resolveRunner(existing: String?) -> String? {
        if let existing, FileManager.default.isExecutableFile(atPath: existing) {
            return existing
        }
        if let bundled = Bundle.main.url(forResource: "session-ping", withExtension: "sh") {
            // The exec bit can get lost in copies; restore it best-effort.
            try? FileManager.default.setAttributes(
                [.posixPermissions: 0o755], ofItemAtPath: bundled.path)
            return bundled.path
        }
        return nil
    }

    /// Apply the desired state: write + (re)load the agent, or unload + remove
    /// it. Returns an error line for the Settings UI, nil on success. Same
    /// swallow-and-surface style as LoginItem: the UI just reflects `read()`.
    @discardableResult
    static func apply(enabled: Bool, schedule: SessionPingSchedule) -> String? {
        if !enabled {
            bootout()
            try? FileManager.default.removeItem(at: plistURL)
            return nil
        }
        guard schedule.isValid else {
            return "Invalid schedule: needs at least one HH:MM time and one weekday."
        }
        guard let runner = resolveRunner(existing: read().runner) else {
            return "session-ping.sh not found - reinstall the app (./install.sh macos)."
        }
        let xml = SessionPingAgent.plistXML(schedule: schedule, runner: runner)
        do {
            try FileManager.default.createDirectory(
                at: plistURL.deletingLastPathComponent(), withIntermediateDirectories: true)
            try xml.write(to: plistURL, atomically: true, encoding: .utf8)
        } catch {
            return "Could not write \(plistURL.path): \(error.localizedDescription)"
        }
        // Reload: bootout then bootstrap, `load -w` fallback for older macOS -
        // the same idiom as install.sh's launchd branch.
        bootout()
        if !launchctl(["bootstrap", "gui/\(getuid())", plistURL.path]) {
            if !launchctl(["load", "-w", plistURL.path]) {
                return "launchctl could not load the agent - see Console.app."
            }
        }
        return nil
    }

    private static func bootout() {
        _ = launchctl(["bootout", "gui/\(getuid())/\(SessionPingAgent.label)"])
    }

    private static func launchctl(_ args: [String]) -> Bool {
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: "/bin/launchctl")
        proc.arguments = args
        proc.standardOutput = Pipe()
        proc.standardError = Pipe()
        guard (try? proc.run()) != nil else { return false }
        proc.waitUntilExit()
        return proc.terminationStatus == 0
    }
}

File: claude-usage-panel@fschmutz.github.io/lib/sessionPing.js
// Session pings from the extension: reads and writes the same systemd user
// units as `./install.sh sessionping`, so the installer and the preferences UI
// stay two frontends over one schedule. The unit files on disk are the source
// of truth - nothing is mirrored into GSettings - exactly like the macOS app
// and its launchd agent.

import GLib from 'gi://GLib';
import Gio from 'gi://Gio';

import {readText, writeText} from './fs.js';
import {stateDir} from './paths.js';
import {run} from './proc.js';

import {
    SP_UNIT, DEFAULT_DAYS, DEFAULT_TIMES,
    parseServiceExec, parseTimerTimes, serviceText, timerText,
} from './sessionPingUnit.js';

function unitDir() {
    return GLib.build_filenamev([GLib.get_user_config_dir(), 'systemd', 'user']);
}

function unitPath(ext) {
    return GLib.build_filenamev([unitDir(), `${SP_UNIT}.${ext}`]);
}

/** Where scripts/session-ping.sh records its last successful ping. */
export function lastPingPath() {
    return GLib.build_filenamev([stateDir(), 'last-ping']);
}

/** The raw stamp written by the last successful ping, or null. */
export function readLastPing() {
    const raw = readText(lastPingPath());
    return raw ? raw.trim() : null;
}

/**
 * The installed schedule.
 * @returns {{enabled: boolean, times: string[], days: number[], runner: ?string}}
 */
export function readSchedule() {
    const timer = readText(unitPath('timer'));
    const {runner, days} = parseServiceExec(readText(unitPath('service')));
    if (!timer)
        return {enabled: false, times: DEFAULT_TIMES.slice(), days: DEFAULT_DAYS.slice(), runner};
    const times = parseTimerTimes(timer);
    return {
        enabled: times.length > 0,
        times: times.length ? times : DEFAULT_TIMES.slice(),
        days: days ?? DEFAULT_DAYS.slice(),
        runner,
    };
}

/**
 * The ping worker script. Prefer the path the installed unit already uses (a
 * git-checkout install), else the copy `install.sh gnome` drops beside the
 * extension. Same resolution order as the macOS app's bundled Resources copy.
 */
export function resolveRunner(existing, extensionPath) {
    if (existing && GLib.file_test(existing, GLib.FileTest.IS_EXECUTABLE))
        return existing;
    const bundled = GLib.build_filenamev([extensionPath, 'scripts', 'session-ping.sh']);
    if (GLib.file_test(bundled, GLib.FileTest.EXISTS)) {
        // The exec bit can be lost in a copy; restore it best-effort.
        try {
            Gio.File.new_for_path(bundled).set_attribute_uint32(
                'unix::mode', 0o755, Gio.FileQueryInfoFlags.NONE, null);
        } catch {
            // bash <runner> still works without it.
        }
        return bundled;
    }
    return null;
}

// Resolves to an error line, or null on success.
async function systemctl(args) {
    const {ok, stderr} = await run(['systemctl', '--user', ...args]);
    return ok ? null : stderr.trim() || 'systemctl failed';
}

export function hasSystemd() {
    return GLib.find_program_in_path('systemctl') !== null &&
        GLib.file_test('/run/systemd/system', GLib.FileTest.IS_DIR);
}

/**
 * Apply the desired schedule: write + enable the units, or disable + remove
 * them. Resolves to an error line for the UI, or null on success.
 */
export async function applySchedule({enabled, times, days, extensionPath}) {
    if (!hasSystemd()) {
        return 'No systemd user session here - schedule pings with ' +
            './install.sh sessionping instead.';
    }
    if (!enabled) {
        await systemctl(['disable', '--now', `${SP_UNIT}.timer`]);
        for (const ext of ['timer', 'service'])
            GLib.unlink(unitPath(ext));
        await systemctl(['daemon-reload']);
        return null;
    }
    if (!times.length)
        return 'Add at least one ping time (HH:MM).';
    const runner = resolveRunner(readSchedule().runner, extensionPath);
    if (!runner)
        return 'session-ping.sh not found - reinstall with ./install.sh gnome.';
    try {
        writeText(unitPath('service'), serviceText(runner, days));
        writeText(unitPath('timer'), timerText(times));
    } catch (e) {
        return `Could not write the systemd units: ${e.message}`;
    }
    const reload = await systemctl(['daemon-reload']);
    if (reload)
        return reload;
    return systemctl(['enable', '--now', `${SP_UNIT}.timer`]);
}
